ProjectBuildLogAppender not found when starting the daemon

Reproducible via `mvnd -v`. The following is looged on the server:

```
09:23:23,083 |-ERROR in ch.qos.logback.core.joran.action.AppenderAction - Could not create an Appender of type [org.jboss.fuse.mvnd.logging.smart.ProjectBuildLogAppender]. ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.jboss.fuse.mvnd.logging.smart.ProjectBuildLogAppender
	at ch.qos.logback.core.util.DynamicClassLoadingException: Failed to instantiate type org.jboss.fuse.mvnd.logging.smart.ProjectBuildLogAppender
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassNameAndParameter(OptionHelper.java:69)
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassName(OptionHelper.java:45)
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassName(OptionHelper.java:34)
	at 	at ch.qos.logback.core.joran.action.AppenderAction.begin(AppenderAction.java:52)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.callBeginAction(Interpreter.java:269)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.startElement(Interpreter.java:145)
	at 	at ch.qos.logback.core.joran.spi.Interpreter.startElement(Interpreter.java:128)
	at 	at ch.qos.logback.core.joran.spi.EventPlayer.play(EventPlayer.java:50)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:165)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:152)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:110)
	at 	at ch.qos.logback.core.joran.GenericConfigurator.doConfigure(GenericConfigurator.java:53)
	at 	at ch.qos.logback.classic.util.ContextInitializer.configureByResource(ContextInitializer.java:75)
	at 	at ch.qos.logback.classic.util.ContextInitializer.autoConfig(ContextInitializer.java:150)
	at 	at org.slf4j.impl.StaticLoggerBinder.init(StaticLoggerBinder.java:84)
	at 	at org.slf4j.impl.StaticLoggerBinder.<clinit>(StaticLoggerBinder.java:55)
	at 	at org.slf4j.LoggerFactory.bind(LoggerFactory.java:150)
	at 	at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:124)
	at 	at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:412)
	at 	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:357)
	at 	at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:383)
	at 	at org.jboss.fuse.mvnd.common.Environment.<clinit>(Environment.java:57)
	at 	at org.jboss.fuse.mvnd.common.MavenDaemon.main(MavenDaemon.java:33)
Caused by: java.lang.ClassNotFoundException: org.jboss.fuse.mvnd.logging.smart.ProjectBuildLogAppender
	at 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
	at 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at 	at ch.qos.logback.core.util.OptionHelper.instantiateByClassNameAndParameter(OptionHelper.java:56)
	at 	... 22 common frames omitted
```

The stack trace was mentioned in https://github.com/mvndaemon/mvnd/issues/152 which seems to be about another problem.

--- a/client/src/main/java/org/jboss/fuse/mvnd/client/DaemonConnector.java
+++ b/client/src/main/java/org/jboss/fuse/mvnd/client/DaemonConnector.java
@@ -16,12 +16,10 @@
 package org.jboss.fuse.mvnd.client;
 
 import java.io.File;
-import java.io.IOException;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.Socket;
 import java.nio.channels.SocketChannel;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -33,9 +31,7 @@ import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
-import java.util.function.Predicate;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
 import org.jboss.fuse.mvnd.common.BuildProperties;
 import org.jboss.fuse.mvnd.common.DaemonCompatibilitySpec;
 import org.jboss.fuse.mvnd.common.DaemonCompatibilitySpec.Result;
@@ -252,11 +248,8 @@ public class DaemonConnector {
         final Path workingDir = layout.userDir();
         String command = "";
         try {
-            String classpath = findJars(
-                    mavenHome,
-                    p -> p.getFileName().toString().equals("mvnd-common-" + buildProperties.getVersion() + ".jar"),
-                    p -> p.getFileName().toString().startsWith("slf4j-api-"),
-                    p -> p.getFileName().toString().startsWith("logback-"));
+            final String classpath = mavenHome.resolve("mvn/lib/ext/mvnd-common-" + buildProperties.getVersion() + ".jar")
+                    .toString();
             final String java = Os.current().isUnixLike() ? "bin/java" : "bin\\java.exe";
             List<String> args = new ArrayList<>();
             args.add(layout.javaHome().resolve(java).toString());
@@ -269,6 +262,9 @@ public class DaemonConnector {
             args.add("-Dmvnd.java.home=" + layout.javaHome().toString());
             args.add("-Dlogback.configurationFile=" + layout.getLogbackConfigurationPath());
             args.add("-Ddaemon.uid=" + uid);
+            if (Boolean.getBoolean(Environment.DEBUG_ENVIRONMENT_PROP)) {
+                args.add("-D" + Environment.DEBUG_ENVIRONMENT_PROP + "=true");
+            }
             args.add("-Xmx4g");
             args.add(Environment.DAEMON_IDLE_TIMEOUT_MS.asCommandLineProperty(Integer.toString(layout.getIdleTimeoutMs())));
             args.add(Environment.DAEMON_KEEP_ALIVE_MS.asCommandLineProperty(Integer.toString(layout.getKeepAliveMs())));
@@ -289,18 +285,6 @@ public class DaemonConnector {
                     String.format("Error starting daemon: uid = %s, workingDir = %s, daemonArgs: %s",
                             uid, workingDir, command),
                     e);
-        }
-    }
-
-    private String findJars(Path mavenHome, Predicate<Path>... filters) {
-        final Path libExtDir = mavenHome.resolve("mvn/lib/ext");
-        try (Stream<Path> jars = Files.list(libExtDir)) {
-            return jars
-                    .filter(Stream.of(filters).reduce((previous, current) -> previous.or(current)).get())
-                    .map(Path::toString)
-                    .collect(Collectors.joining(File.pathSeparator));
-        } catch (IOException e) {
-            throw new RuntimeException("Could not list " + libExtDir);
         }
     }
 

--- a/common/src/main/java/org/jboss/fuse/mvnd/common/Environment.java
+++ b/common/src/main/java/org/jboss/fuse/mvnd/common/Environment.java
@@ -24,6 +24,7 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Properties;
 import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
 import org.slf4j.Logger;
@@ -54,7 +55,33 @@ public enum Environment {
 
     public static final int DEFAULT_MIN_THREADS = 1;
 
-    private static final Logger LOG = LoggerFactory.getLogger(Environment.class);
+    private static final Consumer<String> LOG;
+    private static final boolean DEBUG_ENABLED;
+    public static final String DEBUG_ENVIRONMENT_PROP = "mvnd.environment.debug";
+
+    static {
+        Consumer<String> log = null;
+        boolean debugEnabled = false;
+        try {
+            Logger logger = LoggerFactory.getLogger(Environment.class);
+            log = logger::debug;
+            debugEnabled = logger.isDebugEnabled();
+        } catch (java.lang.NoClassDefFoundError e) {
+            if (e.getMessage().contains("org/slf4j/LoggerFactory")) {
+                /* This is when we are in the daemon's boot class path where slf4j is not available */
+                if (Boolean.getBoolean(DEBUG_ENVIRONMENT_PROP)) {
+                    log = s -> System.out.println("mvnd.environment: " + s);
+                    debugEnabled = true;
+                }
+            } else {
+                throw e;
+            }
+        }
+        LOG = log != null ? log : s -> {
+        };
+        DEBUG_ENABLED = debugEnabled;
+    }
+
     static Properties properties = System.getProperties();
     static Map<String, String> env = System.getenv();
 
@@ -276,7 +303,7 @@ public enum Environment {
                 }
             }
             final String result = valueSource.valueSupplier.get();
-            if (result != null && LOG.isDebugEnabled()) {
+            if (result != null && DEBUG_ENABLED) {
                 StringBuilder sb = new StringBuilder("Loaded environment value for key [")
                         .append(envKey.name())
                         .append("] from ");
@@ -284,7 +311,7 @@ public enum Environment {
                 sb.append(": [")
                         .append(result)
                         .append(']');
-                LOG.debug(sb.toString());
+                LOG.accept(sb.toString());
             }
             return result;
         }

